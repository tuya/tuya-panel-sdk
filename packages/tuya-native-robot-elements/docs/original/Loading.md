# Loading

## 版本

## 代码演示

### 基本用法

```tsx
  <Loading />
```
