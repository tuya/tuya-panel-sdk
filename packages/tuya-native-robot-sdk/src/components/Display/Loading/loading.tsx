import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import PropTypes from 'prop-types';
import { LoadingDoubleBounce, LoadingBubbles, LoadingPulse } from './circle-loading-view';
import CircleLoading from './circle-loading';

/**
 * 把所有load的样式组件集合在一起
 */

const typeEnum = {
  none: 'none',
  doubleBounce: 'double-bounce',
  bubbles: 'bubbles',
  pulse: 'pulse',
  ring: 'ring'
};

export default class Loading extends Component {
  static propTypes = {
    /**
     * loading样式
     */
    type: PropTypes.oneOf([typeEnum.doubleBounce, typeEnum.bubbles, typeEnum.pulse, typeEnum.none]),
    /**
     * render自定义组件
     */
    renderLoading: PropTypes.func,
    /**
     * 大小
     */
    size: PropTypes.number,
    /**
     * 颜色
     */
    color: PropTypes.string,
    /**
     * 容器样式
     */
    style: PropTypes.object,
  };

  static defaultProps = {
    type: typeEnum.bubbles,
    renderLoading: undefined,
    size: 10,
    color: '#eee',
    style: {},
  };

  renderLoadingItem() {
    const { type, size, color } = this.props;
    switch (type) {
      case typeEnum.doubleBounce:
        return <LoadingDoubleBounce size={size} color={color} />;
      case typeEnum.bubbles:
        return <LoadingBubbles size={size} color={color} />;
      case typeEnum.pulse:
        return <LoadingPulse size={size} color={color} />;
      case typeEnum.ring:
        return <CircleLoading />;
      default:
        return <View />;
    }
  }

  render() {
    const { type, renderLoading, style, loadingStyle } = this.props;
    if (renderLoading) return renderLoading();
    return (
      <View style={[styles.root, style]}>
        <View style={[styles.loading, loadingStyle]}>{this.renderLoadingItem()}</View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  root: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  loading: {},
});
