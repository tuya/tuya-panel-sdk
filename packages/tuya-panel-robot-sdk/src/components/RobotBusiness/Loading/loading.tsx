import React, { Component } from 'react';
import { StyleSheet, View, StyleProp, ViewStyle } from 'react-native';
import { LoadingDoubleBounce, LoadingBubbles, LoadingPulse } from './circle-loading-view';
import CircleLoading from './circle-loading';

/**
 * 把所有load的样式组件集合在一起
 */

const typeEnum = {
  none: 'none',
  doubleBounce: 'double-bounce',
  bubbles: 'bubbles',
  pulse: 'pulse',
  ring: 'ring',
};

interface IProps {
  style: StyleProp<ViewStyle>; // 容器样式
  type: string; // loading样式
  size?: number; // 大小
  color?: string; // 颜色
  renderLoading?: () => any; // render自定义组件
  loadingStyle?: StyleProp<ViewStyle>; // loading样式
  spaceBetween?: number;
}

export default class Loading extends Component<IProps> {
  static defaultProps = {
    type: typeEnum.bubbles,
    renderLoading: undefined,
    size: 10,
    color: '#eee',
    style: {},
  };

  renderLoadingItem() {
    const { type, size, color, spaceBetween } = this.props;
    switch (type) {
      case typeEnum.doubleBounce:
        return <LoadingDoubleBounce size={size} color={color} />;
      case typeEnum.bubbles:
        return <LoadingBubbles size={size} color={color} spaceBetween={spaceBetween} />;
      case typeEnum.pulse:
        return <LoadingPulse size={size} color={color} />;
      case typeEnum.ring:
        return <CircleLoading />;
      default:
        return <View />;
    }
  }

  render() {
    const { renderLoading, style, loadingStyle } = this.props;
    if (renderLoading) return renderLoading();
    return (
      <View style={[styles.root, style]}>
        <View style={[styles.loading, loadingStyle]}>{this.renderLoadingItem()}</View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  loading: {},
  root: {
    alignItems: 'center',
    justifyContent: 'center',
  },
});
