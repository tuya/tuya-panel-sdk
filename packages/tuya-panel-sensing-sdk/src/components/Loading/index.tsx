import React from 'react';
import {
  StyleSheet,
  Dimensions,
  Text,
  View,
  Modal,
  ActivityIndicator,
  StyleProp,
  ViewStyle,
  TextStyle,
} from 'react-native';

const SCREEN_WIDTH = Dimensions.get('window').width;
const SCREEN_HEIGHT = Dimensions.get('window').height;

export class EasyLoading {
  static map: Map<any, any>;

  static bind(LoadingThis: any, key: any = 'default') {
    LoadingThis && this.map.set(key, LoadingThis);
  }

  static unBind(key = 'default') {
    this.map.delete(key);
  }

  static show(text = 'Loading...', timeout = -1, key: any = 'default') {
    const result = this.map.get(key);
    result && result.setState({ isShow: true, text, timeout });
  }

  static dismis(text = 'Loading...', key: any = 'default') {
    const result = this.map.get(key);
    result && result.setState({ text, isShow: false });
  }
}

EasyLoading.map = new Map();

export interface LoadingProps {
  type?: string;
  color?: string;
  textStyle?: StyleProp<TextStyle>;
  loadingStyle?: StyleProp<ViewStyle>;
}

export interface LoadingState {
  isShow: boolean;
  timeout: number;
  text: string;
}

export class Loading extends React.Component<LoadingProps, LoadingState> {
  handle = 0;
  constructor(props: LoadingProps) {
    super(props);
    this.state = {
      isShow: false,
      timeout: -1,
      text: 'Loading...',
    };
    EasyLoading.bind(this, this.props.type || 'default');
  }

  componentWillUnmount() {
    clearTimeout(this.handle);
    EasyLoading.unBind(this.props.type || 'default');
  }

  // eslint-disable-next-line react/no-deprecated
  componentWillReceiveProps() {
    EasyLoading.bind(this, this.props.type || 'default'); // 重新绑定
  }

  render() {
    const { type, color, textStyle, loadingStyle } = this.props;
    const { timeout, isShow, text } = this.state;

    clearTimeout(this.handle);

    timeout !== -1 &&
      (this.handle = setTimeout(() => {
        EasyLoading.dismis(type || 'default');
      }, timeout) as any);

    return (
      <Modal
        transparent
        animationType="fade"
        visible={isShow}
        onRequestClose={() => {
          __DEV__ && console.log('Modal has been closed.');
          // alert('Modal has been closed.')
        }}
      >
        <View style={[styles.load_box, loadingStyle]}>
          <ActivityIndicator
            animating
            color={color || '#FFF'}
            size="large"
            style={styles.load_progress}
          />
          <Text style={[styles.load_text, textStyle]}>{text}</Text>
        </View>
      </Modal>
    );
  }
}

const styles = StyleSheet.create({
  load_box: {
    alignItems: 'center',
    backgroundColor: '#0008',
    borderRadius: 10,
    height: 100,
    marginLeft: SCREEN_WIDTH / 2 - 50,
    marginTop: SCREEN_HEIGHT / 2 - 50,
    width: 100,
  },
  load_progress: {
    height: 90,
    position: 'absolute',
    width: 100,
  },
  load_text: {
    color: '#FFF',
    marginTop: 70,
  },
});

export default {
  EasyLoading,
  Loading,
};
