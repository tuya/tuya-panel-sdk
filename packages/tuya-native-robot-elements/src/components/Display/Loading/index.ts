import Loading from './loading';
import { LoadingDoubleBounce, LoadingBubbles, LoadingPulse } from './circle-loading-view';

export {
  LoadingDoubleBounce,
  LoadingBubbles,
  LoadingPulse,
}

export default Loading;
