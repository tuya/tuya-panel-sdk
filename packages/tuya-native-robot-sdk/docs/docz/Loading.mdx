---
name: Loading
nameZh: 通用加载动画组件
menu: tuya-native-robot-elements
route: /tuya-native-robot-elements/Loading
---

import { Playground } from 'docz';
import { TYNative, Utils, I18N, defaultTheme } from '@tuya-rn/tuya-native-components';
import Loading from '@tuya-rn/tuya-native-robot-elements/src/components/loading/loading.tsx';
import Props from '../../Props';

# Loading 通用加载动画组件

## API

<Props path="tuya-native-robot-elements/src/components/loading/loading.tsx" of={Loading} />
