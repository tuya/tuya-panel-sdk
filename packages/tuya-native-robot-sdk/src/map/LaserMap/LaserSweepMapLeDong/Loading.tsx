import React, { PureComponent } from 'react';
import { Image, StyleSheet, View, ActivityIndicator } from 'react-native';
import { Utils, TYText } from '@tuya-rn/tuya-native-components';

import { LoadingBubbles } from '../../../components/Display/Loading';
import I18n from '../../../i18n';
import panelMapConfig, { ILaserMapPanelConfig } from './constant/panelMapConfig';
import res from './res';

const { convertY: cy, viewWidth } = Utils.RatioUtils;

interface IProps {
  loadingShowType?: ILaserMapPanelConfig['map']['loadingShowType'];
  loadingColor?: ILaserMapPanelConfig['map']['loadingColor'];
  loadingImageUri?: ILaserMapPanelConfig['map']['loadingImageUri'];
  isLoading?: boolean;
  isEmpty?: boolean;
  isLite?: boolean; // 小菊花模式
  isShowLoadingTips?: boolean; // 是否展示提示
}
export default class Loading extends PureComponent<IProps> {
  static defaultProps = {
    loadingShowType: panelMapConfig.map.loadingShowType,
    loadingColor: panelMapConfig.map.loadingColor,
    isLoading: false,
    isEmpty: false,
    isShowLoadingTips: false,
  };

  renderTips() {
    const { isLoading, isEmpty, isShowLoadingTips, loadingColor } = this.props;
    const textStyle = [styles.tips, { color: loadingColor }];
    if (!isShowLoadingTips) return null;
    if (isLoading) return <TYText style={textStyle} text={I18n.getLang('mapLoading')} />;
    if (isEmpty) return <TYText style={textStyle} text={I18n.getLang('mapEmpty')} />;
    return null;
  }

  renderLite() {
    const { loadingColor, isEmpty } = this.props;
    if (isEmpty) return null;
    return <ActivityIndicator size="large" color={loadingColor} />;
  }

  renderLoading() {
    const { loadingShowType, loadingColor, loadingImageUri } = this.props;

    // let context;
    switch (loadingShowType) {
      case 'bubble':
        return <LoadingBubbles size={10} color={loadingColor} />;

      case 'image':
        const imageSource = !!loadingImageUri ? { uri: loadingImageUri } : res.mapIcon;
        return <Image source={imageSource} style={styles.image} resizeMode="contain" />;
      default:
        return null;
    }
  }

  render() {
    const { isLite } = this.props;
    return (
      <View style={styles.loading}>
        {isLite ? this.renderLite() : [this.renderLoading(), this.renderTips()]}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  loading: {
    // flex: 1,
    position: 'absolute',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  image: {
    width: cy(120),
    height: cy(120),
    // opacity: 0.8,
  },
  tips: {
    marginTop: cy(20),
    color: 'rgba(255,255,255, 0.5)',
    textAlign: 'center',
    lineHeight: cy(21),
    maxWidth: viewWidth - cy(150),
  },
});
