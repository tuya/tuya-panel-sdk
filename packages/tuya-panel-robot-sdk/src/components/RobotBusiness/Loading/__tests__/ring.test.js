import React from 'react';
import { shallow } from 'enzyme';
import Loading from '../index';

describe('Loading components', () => {
  it('basic render', () => {
    const wrapper = shallow(<Loading type={'ring'} size={20} color={'#ff0987'} />);
    expect(wrapper).toMatchSnapshot();
  });

  // it('componentWillReceiveProps render', () => {
  //   let type = 'double-bounce';
  //   const wrapper = shallow(<Loading type={type} size={20} color={'#ff0987'} />);
  //   type = 'ring';
  //   expect(wrapper).toMatchSnapshot();
  //   wrapper.setProps({ wrapper });
  //   expect(wrapper).toMatchSnapshot();
  //   wrapper.unmount();
  // });
});
