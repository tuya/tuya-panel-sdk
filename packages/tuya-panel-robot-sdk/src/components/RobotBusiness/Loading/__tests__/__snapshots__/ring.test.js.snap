// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading components basic render 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {},
    ]
  }
>
  <View
    style={
      Array [
        Object {},
        undefined,
      ]
    }
  >
    <CircleLoading
      backgroundColor="rgba(0,0,0,.1)"
      color="#fff"
      loading={false}
      size={28}
      strokeWidth={4}
      style={Object {}}
    />
  </View>
</View>
`;
